File: src/io/ziheng/list/leetcode/CopyListWithRandomPointer.java
package io.ziheng.list.leetcode;

import java.util.Map;
import java.util.HashMap;

/**
 * LeetCode 138. Copy List with Random Pointer
 * https://leetcode.com/problems/copy-list-with-random-pointer/
 */
public class CopyListWithRandomPointer {
    private class Node {
        int val;
        Node next;
        Node random;
        public Node(int val) {
            this.val = val;
            this.next = null;
            this.random = null;
        }
    }
    public Node copyRandomList(Node head) {
        if (head == null) {
            return head;
        }
        Map<Node, Node> map = new HashMap<>();
        Node node = head;
        /**
         * Node -> newNode
         */
        while (node != null) {
            map.put(node, new Node(node.val));
            node = node.next;
        }
        node = head;
        while (node != null) {
            map.get(node).next = map.get(node.next);
            map.get(node).random = map.get(node.random);
            node = node.next;
        }
        return map.get(head);
    }
}
/* EOF */

File: src/cc/linkedlist/CopyListWithRandomPointer.cc
#include <cstdlib>
#include <unordered_map>

using namespace std;

/**
 * Definition for a Node.
 */
class Node {
public:
    int val;
    Node *next;
    Node *random;
    Node(int _val): val(_val), next(NULL), random(NULL) {
    }
};

/**
 * LeetCode 138. Copy List with Random Pointer
 * https://leetcode.com/problems/copy-list-with-random-pointer/
 */
class Solution {
public:
    Node* copyRandomList(Node *head) {
        if (head == NULL) {
            return head;
        }
        unordered_map<Node*, Node*> aMap;
        Node *currentNode = head;
        while (currentNode != NULL) {
            aMap.insert(make_pair(currentNode, new Node(currentNode -> val)));
            currentNode = currentNode -> next;
        }
        currentNode = head;
        while (currentNode != NULL) {
            aMap[currentNode] -> next = aMap[currentNode -> next];
            aMap[currentNode] -> random = aMap[currentNode -> random];
            currentNode = currentNode -> next;
        }
        return aMap[head];
    }
};

int main(int argc, char const *argv[]) {
    // ...
    return 0;
}
/* EOF */